src/commonMain/kotlin/app/cash/turbine/Turbine.kt

```diff
@@ -0,0 +1,201 @@
+   *
+   * ```
+   * if (!asChannel().trySend(item).isSuccess) error()
+   * ```
+   */
+  public fun add(item: T)
+
+  /**
+   * Assert that the next event received was non-null and return it.
+   * This function will not suspend, and will throw if invoked in a suspending context.
+   *
+   * @throws AssertionError if the next event was completion or an error.
+   */
+  public fun takeEvent(): Event<T>
+
+  /**
+   * Assert that the next event received was an item and return it.
+   * This function will not suspend, and will throw if invoked in a suspending context.
+   *
+   * @throws AssertionError if the next event was completion or an error, or no event.
+   */
+  public fun takeItem(): T
+
+  /**
+   * Assert that the next event received is [Event.Complete].
+   * This function will not suspend, and will throw if invoked in a suspending context.
+   *
+   * @throws AssertionError if the next event was completion or an error.
+   */
+  public fun takeComplete()
+
+  /**
+   * Assert that the next event received is [Event.Error], and return the error.
+   * This function will not suspend, and will throw if invoked in a suspending context.
+   *
+   * @throws AssertionError if the next event was completion or an error.
+   */
+  public fun takeError(): Throwable
+}
+
+public operator fun <T> Turbine<T>.plusAssign(value: T) { add(value) }
+
+/**
+ * Construct a standalone [Turbine] with an [UNLIMITED] buffer size.
+ */
+public fun <T> Turbine(): Turbine<T> = TurbineImpl()
+
+internal class TurbineImpl<T>(
+  private val channel: Channel<T> = Channel(UNLIMITED),
+  private val job: Job? = null,
+) : Turbine<T> {
+  override fun asChannel(): Channel<T> = object : Channel<T> by channel {
+    override fun tryReceive(): ChannelResult<T> {
+      val result = channel.tryReceive()
+      val event = result.toEvent()
+      if (event is Event.Error || event is Event.Item) _ignoreRemainingEvents = true
+
+      return result
+    }
+
+    override suspend fun receive(): T = try {
+      channel.receive()
+    } catch (e: Throwable) {
+      _ignoreRemainingEvents = true
+      throw e
+    }
+  }
+
+  override fun add(item: T) {
+    if (!asChannel().trySend(item).isSuccess) throw IllegalStateException("Added when closed")
+  }
+
+  override suspend fun cancel() {
+    cancel(null)
+  }
+
+  @OptIn(ExperimentalCoroutinesApi::class)
+  override suspend fun cancel(cause: Throwable?) {
+    if (!channel.isClosedForSend) _ignoreTerminalEvents = true
+    channel.close(cause)
+    job?.cancelAndJoin()
+  }
+
+  override fun takeEvent(): Event<T> = asChannel().takeEvent()
+
+  override fun takeItem(): T = asChannel().takeItem()
+
+  override fun takeComplete() = asChannel().takeComplete()
+
+  override fun takeError(): Throwable = asChannel().takeError()
+
+  private var _ignoreTerminalEvents = false
+  private var _ignoreRemainingEvents = false
+
+  override val ignoreRemainingEvents: Boolean
+    get() = _ignoreRemainingEvents
```

Comment:
```suggestion
  override var ignoreRemainingEvents: Boolean = false
```
It's already an `internal` type, but you can +`private set` to be super strict